src/scportrait/pipeline/extraction.py

```diff
@@ -524,206 +509,145 @@ def _extract_classes(self, px_center, arg, return_failed_ids=False):
+        """Function for writing results to HDF5 file in a thread-safe manner."""
+        with hdf5_lock:
+            with h5py.File(self.output_path, "a") as hf:
+                self._single_cell_data = hf["single_cell_data"]
+                self._single_cell_index = hf["single_cell_index"]
 
-        # convert to correct type
-        cell_ids = cell_ids.astype(self.DEFAULT_SEGMENTATION_DTYPE)
-        cell_ids_removed = cell_ids_removed.astype(self.DEFAULT_SEGMENTATION_DTYPE)
+                for res in results:
+                    save_index, stack, cell_id = res
+                    self._single_cell_data[save_index] = stack
+                    self._single_cell_index[save_index] = [save_index, cell_id]
 
-        self.cell_ids_removed = cell_ids_removed  # save for potentially accessing at later time point
-        self._save_removed_classes(self.cell_ids_removed)
-
-        if self.debug:
-            # visualize some cells for debugging purposes
-            # visualize a random cell for every 100 contained in the dataset
-            n_cells = 100
-            n_cells_to_visualize = len(keep_index) // n_cells
-
-            rng = np.random.default_rng()
-            random_indexes = rng.choice(keep_index, n_cells_to_visualize, replace=False)
-
-            for index in random_indexes:
-                stack = _tmp_single_cell_data[index]
+    def _create_output_files(self):
+        single_cell_index_shape = (self.num_classes, 2)
+        single_cell_data_shape = (
+            self.num_classes,
+            (self.n_masks + self.n_image_channels),
+            self.image_size,
+            self.image_size,
+        )
 
-                fig, axs = plt.subplots(1, stack.shape[0], figsize=(2 * stack.shape[0], 2))
-                for i, img in enumerate(stack):
-                    axs[i].imshow(img, vmin=0, vmax=1)
-                    axs[i].axis("off")
-                fig.tight_layout()
-                plt.show(fig)
+        # generate container for single_cell_data
+        self._tmp_single_cell_data_path = create_empty_mmap(
+            shape=single_cell_data_shape,
+            dtype=np.float16,
+            tmp_dir_abs_path=self._tmp_dir_path,
+        )
 
-        self.log("Transferring extracted single cells to .hdf5")
+        # generate container for single_cell_index
+        fixed_length = 200  # this is the maximum length of string that can be stored in the mmap array for the cell_ids
+        dt = np.dtype(f"S{fixed_length}")
+        self._tmp_single_cell_index_path = create_empty_mmap(
+            shape=single_cell_index_shape,
+            dtype=dt,
+            tmp_dir_abs_path=self._tmp_dir_path,
+        )
 
         # create name for output file
         self.output_path = os.path.join(self.extraction_data_directory, self.DEFAULT_EXTRACTION_FILE)
 
         with h5py.File(self.output_path, "w") as hf:
             hf.create_dataset(
                 "single_cell_index",
```

Comment:
I think these keys should be made constants somewhere like a `constants.py` file so you don't ahve to remember them (like `DEFAULT_EXTRACTION_FILE`)

Reply:
great suggestion. 

Currently they are all gathered together here: https://github.com/MannLabs/scPortrait/blob/15a69689412d7cd73eced2078b9e228bb3d97599/src/scportrait/pipeline/_base.py#L145 

But having this separate would make a lot of sense.

Reply:
I've added them to a separate constants.py file and import them from there throughout the repo. This is much cleaner now as everything is defined in one location.